ionoscloud/firewall_service.go

```diff
@@ -0,0 +1,128 @@
+package ionoscloud
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	ionoscloud "github.com/ionos-cloud/sdk-go/v6"
+	"github.com/ionos-cloud/terraform-provider-ionoscloud/v6/utils"
+	"log"
+	"strconv"
+)
+
+type FirewallService struct {
+	client     *ionoscloud.APIClient
+	meta       interface{}
+	schemaData *schema.ResourceData
+}
+
+func (fs *FirewallService) firewallsGet(ctx context.Context, datacenterId, serverId, nicId string, depth int32) ([]ionoscloud.FirewallRule, error) {
+	firewallRules, apiResponse, err := fs.client.FirewallRulesApi.DatacentersServersNicsFirewallrulesGet(ctx, datacenterId, serverId, nicId).Depth(depth).Execute()
+	logApiRequestTime(apiResponse)
+	if err != nil {
+		return nil, err
+	}
+	if apiResponse.HttpNotFound() || firewallRules.Items == nil || len(*firewallRules.Items) == 0 {
+		log.Printf("[DEBUG] no firewalls found for datacenter %s, server %s, nic %s", datacenterId, serverId, nicId)
+		return nil, nil
+	}
+	return *firewallRules.Items, nil
+}
+
+func (fs *FirewallService) firewallFindById(ctx context.Context, datacenterId, serverId, nicId, firewallId string) (*ionoscloud.FirewallRule, *ionoscloud.APIResponse, error) {
+	firewallRule, apiResponse, err := fs.client.FirewallRulesApi.DatacentersServersNicsFirewallrulesFindById(ctx, datacenterId, serverId, nicId, firewallId).Depth(1).Execute()
+	logApiRequestTime(apiResponse)
+	if err != nil {
+		return nil, apiResponse, err
+	}
+	if apiResponse.HttpNotFound() {
+		log.Printf("[DEBUG] no firewall rule found for datacenter %s, server %s, nic %s", datacenterId, serverId, nicId)
+		return nil, apiResponse, nil
+	}
+	return &firewallRule, apiResponse, nil
+}
+
+func (fs *FirewallService) deleteFirewallRule(ctx context.Context, datacenterId, serverId, nicId, firewallId string) (*ionoscloud.APIResponse, error) {
+	apiResponse, err := fs.client.FirewallRulesApi.DatacentersServersNicsFirewallrulesDelete(ctx, datacenterId, serverId, nicId, firewallId).Execute()
+	apiResponse.LogInfo()
+	if err != nil {
+		return apiResponse, err
+	}
+	// Wait, catching any errors
+	_, errState := getStateChangeConf(fs.meta, fs.schemaData, apiResponse.Header.Get("Location"), schema.TimeoutDelete).WaitForStateContext(ctx)
+	if errState != nil {
+		return apiResponse, fmt.Errorf("an error occured while waiting for state change dcId: %s server_id: %s nic_id %s ID: %s Response: %w", datacenterId, serverId, nicId, firewallId, errState)
```

Comment:
@cristiGuranIonos I would add some commas in this error message: dcId: %s, server_id: %s, nic_id .. and so on.